x/poolmanager/routes.go

```diff
@@ -0,0 +1,569 @@
+package poolmanager
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/osmoutils"
+	cosmwasmpooltypes "github.com/osmosis-labs/osmosis/v20/x/cosmwasmpool/types"
+	gammtypes "github.com/osmosis-labs/osmosis/v20/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v20/x/poolmanager/types"
+)
+
+var (
+	OSMO                 = "uosmo"
+	superfluidMultiplier = sdk.MustNewDecFromStr("1.5")
+	directRouteCache     map[string]uint64
+	spotPriceCache       map[string]osmomath.BigDec
+)
+
+func init() {
+	directRouteCache = make(map[string]uint64)
+	spotPriceCache = make(map[string]osmomath.BigDec)
+}
+
+// findDirectRoute finds all direct routes between two tokens
+func findDirectRoute(g types.RoutingGraphMap, start, end string) []*types.Route {
+	if innerMap, exists := g.Graph[start]; exists {
+		if routes, exists := innerMap.InnerMap[end]; exists {
+			for i := range routes.Routes {
+				routes.Routes[i].Token = end
+			}
+			return routes.Routes
+		}
+	}
+	return nil
+}
+
+// findTwoHopRoute finds all two-hop routes between two tokens
+func findTwoHopRoute(g types.RoutingGraphMap, start, end string) [][]*types.Route {
+	var routeRoutes [][]*types.Route
+
+	if startRoutes, exists := g.Graph[start]; exists {
+		for token, routes := range startRoutes.InnerMap {
+			if tokenRoutes, exists := g.Graph[token]; exists {
+				if endRoutes, exists := tokenRoutes.InnerMap[end]; exists {
+					for _, startRoute := range routes.Routes {
+						startRoute.Token = token
+						for _, endRoute := range endRoutes.Routes {
+							endRoute.Token = end
+							routeRoutes = append(routeRoutes, []*types.Route{startRoute, endRoute})
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return routeRoutes
+}
```

Comment:
nit: consider refactoring to following
```suggestion
func findTwoHopRoute(g types.RoutingGraphMap, start, end string) [][]*types.Route {
	var routeRoutes [][]*types.Route

	startRoutes, startExists := g.Graph[start]
	if !startExists {
		return routeRoutes
	}

	for token, routes := range startRoutes.InnerMap {
		tokenRoutes, tokenExists := g.Graph[token]
		if !tokenExists {
			continue
		}

		endRoutes, endExists := tokenRoutes.InnerMap[end]
		if !endExists {
			continue
		}

		for _, startRoute := range routes.Routes {
			startRoute.Token = token
			for _, endRoute := range endRoutes.Routes {
				endRoute.Token = end
				routeRoutes = append(routeRoutes, []*types.Route{startRoute, endRoute})
			}
		}
	}

	return routeRoutes
}

```